Large refactor

singingwolfboy opened the pull request:
This refactor makes the following changes:
- The crawler is now called with `scrapy crawl edx` (as per Scrapy's documention) instead of the custom `pa11ycrawler` script
- Settings and complexity have been drastically cleaned up, so the whole thing is simpler to read and understand
- Functionality to transform JSON into HTML has been pulled out of the crawler into a separate script (`pa11ycrawler-html`), and also simplified
- The existing tests have been removed, and new ones added to replace them
- README has been updated
- No need to install `pa11y` globally -- this project now uses the local `npm` install
- The crawler will now ignore the following kinds of URLs:
  - URLs where the path starts with `/xblock/`
  - URLs where the path starts with `/api/` (part of Django Rest Framework) -- these pages will be scraped for further links, but not checked for accessibility violations
  - URLs containing `/logout/` (this was there before)
  - URLs containing `?_accept=application/x-tgz`, so the crawler doesn't try to download course exports

This will need to be coordinated with https://github.com/edx/edx-platform/pull/13225

@clrux @cptvitamin do either of you want to do a code review? If you want me to just merge it, I'll need the permissions -- I don't seem to have merge permissions on this repo. :(

singingwolfboy committed:
Large refactor

* The crawler is now called with `scrapy crawl local-edx` (as per Scrapy's documention) instead of the custom `pa11ycrawler` script
* Settings and complexity have been drastically cleaned up, so the whole thing is simpler to read and understand
* Functionality to transform JSON into HTML has been pulled out of the crawler into a separate script, and also simplified
* The existing tests have been removed, and new ones added to replace them
* README has been updated
* No need to install `pa11y` globally -- this project now uses the local `npm` install
* The crawler will now ignore the following kinds of URLs:
  * URLs where the path starts with `/xblock/`
  * URLs where the path starts with `/api/` (part of Django Rest Framework) -- these pages will be scraped for further links, but not checked for accessibility violations
  * URLs containing `/logout/` (this was there before)
  * URLs containing `?_accept=application/x-tgz`, so the crawler doesn't try to download course exports

downzer0 commented:
Nice work! I might ping @jzoldak or @benpatterson on this one.

singingwolfboy committed:
Make crawler more flexible

singingwolfboy committed:
fixed typo

singingwolfboy committed:
Install HTML script as pa11ycrawler-html

benpatterson commented:
@jzoldak can you lend a hand with review? I will take a pass now but will be largely unavailable tomorrow (just a heads up)

benpatterson commented on Makefile:
```diff
@@ -1,24 +1,23 @@
 
 requirements: requirements.js
-	pip install -q -r requirements/base.txt --exists-action w
+	pip install .
```
I haven't seen this before. Can you explain how it works?

singingwolfboy replied:
The `.` is a reference to the current directory. When `pip install` is called with a reference to a directory, it checks for a `setup.py` script inside that directory, and if it finds it, it calls `python setup.py install`, which installs the dependencies specified in the `install_requires` section of the `setup.py` file. However, if you prefer a `requirements.txt` file, I can pull that out and revert this.

benpatterson replied:
I think that's ok; let's just be sure that it can be picked up by gemnasium or VersionEye. Could you check that out?

benpatterson commented on Makefile:
```diff
@@ -1,24 +1,23 @@
 
-develop: requirements.test
+develop: requirements
+	pip install -U pytest>=2.7 pytest-mock pycodestyle pylint edx-lint==0.5.1
```
I'd prefer these be in a requirements file. Updates are easier to follow over time if it's in a file with line breaks (for example)

benpatterson replied:
Same comment here though....Groking your other feedback, I suppose what I'd prefer here is for `develop` to call the test requirements in pip install. I'm assuming that's done through `pip install .` but am less familiar with that convention so I lean on you for the right pattern. Ultimately though I don't want the list of libraries to be on this line, but stored somewhere else. (Which for now is setup.py)

benpatterson commented on Makefile:
```diff
@@ -1,24 +1,23 @@
+	pip install -U pytest>=2.7 pytest-mock pycodestyle pylint edx-lint==0.5.1
 
 clean:
 	find . -name '*.pyc' -delete
```
should this clean out the data, too?

singingwolfboy replied:
The crawler can output data to whatever directory you specify, so there's no way to reliably do this. I can set it to delete the "data" directory if it finds it, though, since that's the default output location.

benpatterson replied:
We should have some easy convention for deleting the crawler results. I think we had a longer discussion on this elsewhere in the PR so I'll comment over there.

benpatterson commented on README.md:
```diff
@@ -1,109 +1,67 @@
-crawling a website and storing a Pa11y report for each page.
-
-The reports can be in the form of any of Pa11y's default report types or [1.0-json](https://github.com/springernature/pa11y-reporter-1.0-json).  There is an additional command to produce html reports from the 1.0-json.
+pa11ycrawler a [Scrapy](http://doc.scrapy.org/en/latest/index.html)
```
nit: pa11ycrawler _is_ a Scrapy...

benpatterson commented on README.md:
```diff
@@ -1,109 +1,67 @@
-
-The reports can be in the form of any of Pa11y's default report types or [1.0-json](https://github.com/springernature/pa11y-reporter-1.0-json).  There is an additional command to produce html reports from the 1.0-json.
+pa11ycrawler a [Scrapy](http://doc.scrapy.org/en/latest/index.html)
+spider that calls [Pa11y](http://pa11y.org/) on every page of a locally-running
```
nit: "that runs a Pa11y check on every page..."

benpatterson commented on README.md:
````diff
@@ -1,109 +1,67 @@
-Install node Requirements
--------------------------
-```
-npm install -g pa11y@3.6.0 pa11y-reporter-1.0-json@1.0.2
````
right-o; i had missed the global install previously. thanks for cleaning that up.

benpatterson commented on README.md:
```diff
@@ -1,109 +1,67 @@
+This data directory is "data" by default, which means it will create a directory
+named "data" in whatever directory you run the crawler from.
+Whatever directory you specify, it will be automatically created if it does
+not yet exist. In addition, this tool will never delete data from the data
```
ah i see this now. i'm curious your thinking here....i would rather be able to clean things up (for convenience, and for ensuring clean automation, etc).

singingwolfboy replied:
In general, I think it makes sense to have "creating data" and "deleting data" be two explicitly different tasks, and running one should never cause the other to happen. If I intend to create data, and I end up deleting data by accident, that could result in a lot of wasted time and effort (in the best case) or catastrophic data loss (in the worst case). For this specific tool, the data can be easily re-created, so deleting it by accident is not usually a bad thing. However, what if the user calling this tool accidentally specified an data directory that had some very important files in it? If the tool deleted the directory before running, you could end up in a very bad situation.

I'm not against having an explicit "clean" or "delete" task, but I really don't think it should be automatically folded into the "crawl" task unless it's done at a higher level (like using Paver to script this).

benpatterson replied:
There's got to be some way of programmatically cleaning out your results. I had asked about this in the `clean` task, and you pointed out that we can't specify a data directory. Here you point out that we shouldn't clean it as a natural part of a run. But we must have a way of clearing out old results, and that should be any one of the following: a make task, a pre-run task, a specific clean-my-data-dir task. I can understand your arguments and will not push too hard on the specific examples; however, we must have a way of cleaning out the data through either the Makefile or the crawl tasks. At the least it will be much more convenient and at the most we can have it scripted as part of any sort of cron job.

benpatterson commented on README.md:
````diff
@@ -1,109 +1,67 @@
-```
-make quality
-```
+This project has tests for the pipeline functions, where are where the main
````
nit: typo towards the end of this line

benpatterson commented on pa11ycrawler/pipelines.py:
```diff
@@ -20,89 +21,150 @@ class DuplicatesPipeline(object):
+    Runs the Pa11y CLI against `item['url']`, using the same request headers
+    used by Scrapy.
+    """
+    pa11y_path = "node_modules/.bin/pa11y"
```
Will we raise if pa11y isn't found? (I'm thinking in this case if someone installs it globally, even though we didn't expect them to)

singingwolfboy replied:
I added a commit to make the `Pa11yPipeline` check to see if `pa11y` is installed when the pipeline is instantiated. If not, it raises a `scrapy.exceptions.NotConfigured` exception.

benpatterson replied:
right on; thanks

benpatterson commented on pa11ycrawler/pipelines.py:
```diff
@@ -20,89 +21,150 @@ class DuplicatesPipeline(object):
+        # but that gets complicated (long URLs, special characters, etc)
+        # so we'll make the filename a hash of the URL instead,
+        # and throw in the access time so that we can store the same URL
+        # multiple times in this data directory
```
There was an original implementation that went along these lines, too, but we had found that troubleshooting was much easier if we could more-easily connect the results with the particular json file. How will we want to do that if we need to look at the original data from, say, the html report, or console output?

singingwolfboy replied:
The existing code (before this pull request) uses hashes for filenames, as well. Can you suggest an alternate implementation that would work better for this?

benpatterson replied:
No, I just mis-read your description. When I heard 'hash' I was thinking md5 hash only. But reading the code more closely I see you're adding to that as a starting point. That's fine by me.

benpatterson commented on pa11ycrawler/pipelines.py:
```diff
@@ -20,89 +21,150 @@ class DuplicatesPipeline(object):
+                # https://www.npmjs.com/package/truffler
+                # If this is the error, we can resolve it just by trying again,
+                # so decrement the retries_remaining and start over.
+                retries_remaining -= 1
```
nice idea with a retry

singingwolfboy replied:
Thanks! 😄

benpatterson commented on pa11ycrawler/settings.py:
```diff
@@ -101,5 +19,15 @@
 }
+
+# Other items you are likely to want to override ---------------
+CONCURRENT_REQUESTS = 16
```
Oo multithreading? Will this help us with the speed of the audit?

singingwolfboy replied:
Concurrency is built into Scrapy. Relevant settings are here: http://doc.scrapy.org/en/1.1/topics/settings.html#concurrent-items

This item was already in the `settings.py` file, I just moved things around so Git is a bit confused and is marking it as a new addition.

benpatterson replied:
Ok great. I wonder if we could do some optimizations here, then, so we can shorten it to the point where it can go into PR builds. What do you think? (Obviously separate from this PR, but...)

benpatterson commented on pa11ycrawler/spiders/edx.py:
```diff
@@ -0,0 +1,94 @@
+
+        # set start URL based on course_key, which is the test course by default
+        api_url = (
+            URLObject("http://")
```
will we be able to crawl edx.org via `http`?

singingwolfboy replied:
This is just the starting URL. On production, `http://` URLs are redirected to `https://`, and the crawler will follow those redirects appropriately. When running on localhost, setting up `https://` is a pain, so it's nicer to start with `http://` instead.

benpatterson replied:
Gotcha. Cool.

benpatterson commented on tests/test_pipelines.py:
```diff
@@ -0,0 +1,115 @@
+
+    # third is the same as a previous, so raises an exception
+    item3 = {"url": "google.com"}
+    with pytest.raises(DropItem):
```
a bit of a testing nit, but can you assert that, also, it does not crawl the duplicate? (depends on how the raise is handled, for example)

singingwolfboy replied:
Unfortunately, I can't. This is due to the architecture of Scrapy, and probably requires some explanation.

We're using three different Scrapy concepts here: spiders, items, and pipelines. A spider's job is to start from a given URL (or list or URLs) and crawl links to get from page to page. Each page is parsed into an item, which represents structured data from the page that a pipeline can operate on. In this case, `EdxSpider` is crawling pages and turning each one into an `A11yItem`, which saves the URL, request headers, access time, and page title. Spiders continue crawling until they've hit every URL they can, or until they're given a signal to stop.

The items produced by spiders are put into a queue and processed by the pipelines. In this case, the `Pa11yPipeline` takes an item, runs `pa11y` with the given URL and headers, and saves the result into a file with the page title and access time. Spiders and pipelines operate completely independently of each other, and by the time a pipeline has started processing an item, the spider that produced the item has already moved on and started crawling other pages.

So because of this architecture, a pipeline can't tell a spider not to crawl a page. The most we can do is drop items that are duplicates of pages we've seen before, so we don't waste effort re-running `pa11y` needlessly.

benpatterson replied:
Ah ok, I get it. Thanks for explaining.

benpatterson replied:
Ok reading through this more, perhaps what I"m asking is for you to ensure process_item doesn't return an item. So, `item3 <> processed3`?

benpatterson commented on tests/test_pipelines.py:
```diff
@@ -0,0 +1,115 @@
+    # non-api url is fine, though
+    item = {"url": "http://courses.edx.org/course/whatever"}
+    processed = drf_pl.process_item(item, spider)
+    assert item == processed
```
I'm a bit less familiar with pytest, I admit. Is there a way to configure a message so a test failure can be explained?

singingwolfboy replied:
This is a regular [`assert` statement](https://docs.python.org/2/reference/simple_stmts.html#grammar-token-assert_stmt) in Python. If you want a customize the failure message, you can do so with a second expression, like this:

``` python
assert 1 == 2, "One is not equal to two"
```

However, I've found that test failures are pretty easy to debug with pytest even without custom failure messages, because of the way that pytest can introspect variables and display their output to you in the failing test. [Here's an example of that.](http://docs.pytest.org/en/latest/example/reportingdemo.html)

benpatterson replied:
Yeah really my problem is with using assert; I've seen many tests where assert x == y wherein it is unclear what x or y are. It still takes some extra winding through the code to figure that out here, but I might've been giving you a knee-jerk reaction over using assert.

benpatterson commented:
Done with my pass. 

I do have some questions that I'd appreciate your feedback on.

In general, nothing particularly stuck out to me; the changes make sense on my first read of them. I haven't looked at the corresponding job output yet, though. I defer to @cptvitamin on whether or not that will be better for him. Also it's worth noting that we have a splunk report that looks at console output to find our pa11y count trends. I wonder if that will need to be changed...could you look at that @singingwolfboy ? I only wrote it recently; so we don't need to refactor the crawler to make it put in the same output; moreso I'd want to refactor the splunk report so it keeps giving us trends. (If it needs to be refactored, that is.)

For the test file, I think you should also add testcases for whatever pa11y returns. We should be able to handle errors, and success, for example.  I realize it's outside the bounds of the tool (e.g., it's a dependency), but it would probably help us out in the long run if we had some end-to-end test, so that whenever the pa11y node library is updated, it's giving us the same results we expect. I don't have a quick answer for you on that, but I'd like to hear your ideas.

Thanks for putting this together; I'll stand by for feedback/comments. I'll be offline much of tomorrow but will try to reply if I can.

singingwolfboy committed:
Correct README typos

singingwolfboy committed:
Pa11yPipeline checks that pa11y is installed

singingwolfboy committed:
Move dependencies out to requirements files

singingwolfboy commented:
> For the test file, I think you should also add testcases for whatever pa11y returns. We should be able to handle errors, and success, for example. I realize it's outside the bounds of the tool (e.g., it's a dependency), but it would probably help us out in the long run if we had some end-to-end test, so that whenever the pa11y node library is updated, it's giving us the same results we expect. I don't have a quick answer for you on that, but I'd like to hear your ideas.

I understand your rationale here, but I don't think it's necessary. First of all, pa11y's return format is totally dependent on which formatter you choose to use. Right now, it's using the [1.0-json](https://www.npmjs.com/package/pa11y-reporter-1.0-json) reporter, which exists primarily to support tools that are dependent on the older output format. It would be nice to be able to upgrade to the newer JSON output format, which has a bit more information. (It just didn't seem worthwhile to do that in this pull request, which is already plenty large enough.)

Second, there is a cost to writing tests. Those tests must be maintained, understood, upgraded, etc, and that puts a tax on new development. For something like this, errors are easy to spot (since `pa11ycrawler-html` script won't run properly), and should be easy to revert. I don't really feel it's necessary.

That being said, if you tell me to write more tests for this, I can do that -- I'm just giving you my thoughts and perspective here. Let me know what you'd like.

singingwolfboy committed:
Need to install code before tests can run

singingwolfboy committed:
Cleaned up Makefile a bit

benpatterson commented:
@singingwolfboy I'm done with my second pass.

The only must for me at this point is the data clean-up. I'm not as concerned with the implementation (I understand your arguments), but my main issue here is having some way of programmatically cleaning out the data. 

I also think a couple explicit asserts on the dropped items would be good, but admittedly is not a must for me.

singingwolfboy committed:
Change default data/html location to /var/opt/pa11ycrawler, add `make clean` task

singingwolfboy commented:
@benpatterson I added a commit that changes the default data directory to `/var/opt/pa11ycrawler`, and updated the `make clean` task to run `rm -rf /var/opt/pa11ycrawler`. Does that do what you want?

benpatterson commented:
Thanks @singingwolfboy. That's a step in the right direction. I had imagined it living under the project directory somewhere. (Again in that case, I'm picturing what would happen if we ran it in automation and had multiple instances running against different sites.) Could you change it to something within the project directory? Perhaps instead of `data/` it could be `data/reports/pa11ycrawler/` Something like that? If you want to have a different `make` target for the report output vs the *.pyc files, that would be fine IMO.

singingwolfboy committed:
Switch back to local data and html directories

benpatterson commented:
Last change looks good to me @singingwolfboy 

Overall I give 👍 assuming tests pass

benpatterson commented:
I say we're good to merge. @jzoldak if you end up reviewing and giving feedback, I suggest we put that in a separate, follow-up PR.

singingwolfboy committed:
Increment project version

cptvitamin commented:
👍

jzoldak commented:
@singingwolfboy @benpatterson Thx for your patience, and apologies for not getting to this PR sooner. LGTM 👍
